crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1601 @@
+                        .is_none(),
+                    "should have already been shut down"
+                );
+            } else if send[0] == PortState::Done {
+                {
+                    // Early shutdown - our port state was set to `Done` by the downstream nodes. This
+                    // also means we can discard any potential errors from the `shutdown()`. Note that
+                    // `shutdown()` internally unwraps any critical errors.
+                    let _ = polars_io::pl_async::get_runtime()
+                        .block_on(self.try_shutdown())
+                        .unwrap();
+                };
+                self.is_finished.store(true, Ordering::Relaxed);
+            } else {
+                send[0] = PortState::Ready
+            }
+        }
+
+        fn spawn<'env, 's>(
+            &'env mut self,
+            scope: &'s TaskScope<'s, 'env>,
+            recv: &mut [Option<RecvPort<'_>>],
+            send: &mut [Option<SendPort<'_>>],
+            _state: &'s ExecutionState,
+            join_handles: &mut Vec<JoinHandle<PolarsResult<()>>>,
+        ) {
+            use std::sync::atomic::Ordering;
+
+            debug_assert!(recv.is_empty());
+            debug_assert_eq!(send.len(), 1);
+
+            let morsel_senders = send[0].take().unwrap().parallel();
+
+            #[cfg(debug_assertions)]
+            {
+                let guard = self.row_group_data_rx_handles.try_read().unwrap();
+                let (row_group_data_receivers, _) = guard.as_ref().unwrap();
+                assert_eq!(row_group_data_receivers.len(), morsel_senders.len());
+            }
+
+            self.row_group_data_task_init_permit.take();
+            let row_group_decoder = self.row_group_decoder.as_ref().unwrap().clone();
+
+            for (i, mut morsel_tx) in morsel_senders.into_iter().enumerate() {
+                let this = &*self;
+                let row_group_decoder = row_group_decoder.clone();
+                let row_group_data_rx_handles_guard =
+                    this.row_group_data_rx_handles.try_read().unwrap();
+
+                join_handles.push(scope.spawn_task(TaskPriority::Low, async move {
+                    let (row_group_data_receivers, _) =
+                        row_group_data_rx_handles_guard.as_ref().unwrap();
+                    let row_group_data_rx = &row_group_data_receivers[i];
+
+                    let wait_group = WaitGroup::default();
+                    let source_token = SourceToken::new();
+
+                    loop {
+                        let Ok(v) = row_group_data_rx.recv().await else {
+                            this.is_finished.store(true, Ordering::Relaxed);
+                            drop(row_group_data_rx_handles_guard);
+                            // Shutdown to propagate any errors.
+                            if let Some(v) = this.try_shutdown().await {
+                                v?;
+                            };
```

Comment:
Instead of this shutdown being part of the main loop with a flag, I would suggest collecting all tasks in a vec and joining on them. Once all tasks have joined, then do the shutdown. Then you're also guaranteed to have a single shutdown, and error propagation becomes trivial.

For an example of this pattern see `nodes/ordered_union.rs`, note how it collects all handles in `inner_handles`, and awaits all of them in a second task.

Reply:
Ah - that's a neat trick, I've changed it to do that. It also lets me avoid doing `Mutex<Receiver>` 👍